### PR TITLE
new MetricNamespace and MetricContext

### PR DESCRIPTION
**Pull request by DanSimon**
This PR solves a lot of issues we've faced in metrics with relative paths.  Now there is a new `MetricNamespace` that is required when creating new metrics.  `MetricSystem` implements this trait and can be used as a namespace, but now we have the ability to create sub-spaces with extended prefixes.

This is now used by all components of Colossus that create their own metrics.

### CODE REVIEW

File: colossus-examples/src/main/scala/colossus-examples/WebsocketExample.scala [hunk 1]
```diff
@@ -52,10 +52,10 @@ object WebsocketExample {
     
     val generator = io.actorSystem.actorOf(Props[PrimeGenerator])
 
-    Server.basic("websocket", port){ new Service[Http](_) {
+    Server.basic("websocket", port){ ctx => new Service[Http](ctx) {
       def handle = {
         case UpgradeRequest(resp) => {
-          become(new WebsocketHandler(_) with ProxyActor {
+          become(_ => new WebsocketServerHandler(ctx) with ProxyActor {
```
Comment (DanSimon) [inline colossus-examples/src/main/scala/colossus-examples/WebsocketExample.scala]: I've created #359 to address the fact that the input context is now not used here.

File: colossus-tests/src/test/scala/colossus/core/ServerSpec.scala [hunk 2]
```diff
@@ -236,7 +236,8 @@ class ServerSpec extends ColossusSpec {
         }
       }
 
-      "open up spot when connection closes" in {
+      //TODO : fix this test
```
Comment (DanSimon) [inline colossus-tests/src/test/scala/colossus/core/ServerSpec.scala]: I've created #360 to address this

File: colossus/src/main/scala/colossus/core/Worker.scala [hunk 3]
```diff
@@ -157,11 +156,10 @@ private[colossus] class Worker(config: WorkerConfig) extends Actor with ActorLog
 
   private val workerIdTag = Map("worker" -> (io.name + "-" + workerId.toString))
 
-  import config.io.metrics.base
-
-  val eventLoops              = Rate(io.namespace / "worker" / "event_loops")
-  val numConnections          = Counter(io.namespace / "worker" / "connections")
-  val rejectedConnections     = Rate(io.namespace / "worker" / "rejected_connections")
+  implicit val ns = io.metrics
```
Comment (nsauro) [inline colossus/src/main/scala/colossus/core/Worker.scala]: I might be mistaken, but isn't this changing the address of these metrics?  Before, they were using the io.namespace, now they are using the MetricSystem namespace, wouldn't that mean they lose any iosystem prefix that was there previously?

Comment (DanSimon) [inline colossus/src/main/scala/colossus/core/Worker.scala]: Ah yeah I forgot the IOSystem needs its own namespace

File: colossus/src/main/scala/colossus/controller/Controller.scala [hunk 4]
```diff
@@ -42,6 +41,8 @@ trait MasterController[Input, Output] extends ConnectionHandler {
   protected def codec: Codec[Output, Input]
   protected def controllerConfig: ControllerConfig
 
+  implicit def namespace : metrics.MetricNamespace
```
Comment (nsauro) [inline colossus/src/main/scala/colossus/controller/Controller.scala]: Not sure for Controller, but shouldn't this also be using an IOSystem's namespace?

File: colossus/src/main/scala/colossus/core/Server.scala [hunk 5]
```diff
@@ -95,6 +95,8 @@ case class ServerRef private[colossus] (config: ServerConfig, server: ActorRef,
 
   def serverState = serverStateAgent.get()
 
+  val namespace : MetricNamespace = system.metrics / name
```
Comment (nsauro) [inline colossus/src/main/scala/colossus/core/Server.scala]: here as well.

File: colossus/src/main/scala/colossus/service/ServiceClient.scala [hunk 6]
```diff
@@ -113,20 +113,20 @@ with ClientConnectionHandler with Sender[P, Callback] with ManualUnbindHandler {
 
   import colossus.core.WorkerCommand._
   import config._
-  import context.worker.metrics
+  implicit val namespace = context.worker.system.metrics / config.name
```
Comment (nsauro) [inline colossus/src/main/scala/colossus/service/ServiceClient.scala]: This actually brings up a question, which I"m not sure I know if we answered or not, which is , how should Client addresses be namespace?  I guess IOSystem?  I can't think of anything lower that would make sense

File: colossus-metrics/src/main/scala/colossus/metrics/collectors/Histogram.scala [hunk 7]
```diff
@@ -97,11 +97,11 @@ object Histogram extends CollectorConfigLoader{
     * @param collection The collection which will contain this Collector.
     * @return
     */
-  def apply(address : MetricAddress, configPath : String)(implicit collection : Collection) : Histogram = {
+  def apply(address : MetricAddress, configPath : String)(implicit ns : MetricNamespace) : Histogram = {
 
     import scala.collection.JavaConversions._
 
-    val params = resolveConfig(collection.config.config, s"$configPath.$address", DefaultConfigPath)
```
Comment (nsauro) [inline colossus-metrics/src/main/scala/colossus/metrics/collectors/Histogram.scala]: config.config makes me sad.  I've bought this upon us, I'll see if i can make it better somehow

File: colossus/src/main/scala/colossus/controller/Controller.scala (hunk 4 above)
Comment (DanSimon) [inline colossus/src/main/scala/colossus/controller/Controller.scala]: This one is abstract, but which namespace it uses depends on whether it's a server or client.

File: colossus/src/main/scala/colossus/service/ServiceClient.scala (hunk 6 above)
Comment (DanSimon) [inline colossus/src/main/scala/colossus/service/ServiceClient.scala]: yeah it should be iosystem.

File: colossus/src/main/scala/colossus/controller/Controller.scala (hunk 4 above)
Comment (nsauro) [inline colossus/src/main/scala/colossus/controller/Controller.scala]: Should it not have anything assigned to it, to force subs to implement it ?

Comment (DanSimon) [inline colossus/src/main/scala/colossus/controller/Controller.scala]: it doesn't.

Comment (nsauro) [inline colossus/src/main/scala/colossus/controller/Controller.scala]: Lol.  wow, I need to learn to read :)

**Comment by DanSimon** [comment]
Fixed the IOSystem namespace issues, fixed conflicts from develop, and fixed the `ClassCastException` issue with `getOrAdd`.

**Comment by nsauro** [comment]
👍

**Pull request merged**